### PR TITLE
core: Add specific signals type hinting

### PR DESCRIPTION
**Pull request by yotamN**
Classes with message listeners are:

- [x] Script
- [x] Session
- [ ] Bus
- [x] Device
- [x] DeviceManager
- [x] PortalService
- [x] Compiler

### CODE REVIEW

File: frida/core.py [hunk 1]
```diff
@@ -578,6 +582,18 @@ def post(self, message: Any, data: Optional[Union[str, bytes]] = None) -> None:
         _filter_missing_kwargs(kwargs)
         self._impl.post(raw_message, **kwargs)
 
+    @overload
+    def on(self, signal: Literal["detached"], callback: Callable[[], None]) -> None:
```
Comment (oriori1703) [inline frida/core.py:586]:
```suggestion
    def on(self, signal: Literal["detached"], callback: Callable[[], Any]) -> None:
```

File: frida/core.py [hunk 2]
```diff
@@ -254,7 +254,11 @@ def disable_debugger(self) -> None:
         self._impl.disable_debugger()
 
     @overload
-    def on(self, signal: Literal["message"], callback: Callable[[Mapping[Any, Any], Any], Any]) -> None:
+    def on(self, signal: Literal["destroyed"], callback: Callable[[], None]) -> None:
```
Comment (oriori1703) [inline frida/core.py:257]: If the return value of a callback is ignored I think it is better to have it has `Any` instead of `None`. 
```suggestion
    def on(self, signal: Literal["destroyed"], callback: Callable[[], Any]) -> None:
```

File: frida/core.py [hunk 3]
```diff
@@ -1087,6 +1103,30 @@ def untag(self, connection_id: int, tag: str) -> None:
...
+    # TODO: add those callbacks
+    # public signal void node_connected (uint connection_id, SocketAddress remote_address);
+    # public signal void node_joined (uint connection_id, Application application);
+    # public signal void node_left (uint connection_id, Application application);
+    # public signal void node_disconnected (uint connection_id, SocketAddress remote_address);
+    # public signal void controller_connected (uint connection_id, SocketAddress remote_address);
+    # public signal void controller_disconnected (uint connection_id, SocketAddress remote_address);
+
+    @overload
+    def on(self, signal: Literal["authenticated"], callback: Callable[[int, str], None]) -> None:
```
Comment (oriori1703) [inline frida/core.py:1115]:
```suggestion
    def on(self, signal: Literal["authenticated"], callback: Callable[[int, str], Any]) -> None:
```

File: frida/core.py [hunk 4]
```diff
@@ -1087,6 +1103,30 @@ def untag(self, connection_id: int, tag: str) -> None:
...
+        ...
+
+    @overload
+    def on(self, signal: Literal["subscribe"], callback: Callable[[int], None]) -> None:
```
Comment (oriori1703) [inline frida/core.py:1119]:
```suggestion
    def on(self, signal: Literal["subscribe"], callback: Callable[[int], Any]) -> None:
```

File: frida/core.py [hunk 5]
```diff
@@ -1087,6 +1103,30 @@ def untag(self, connection_id: int, tag: str) -> None:
...
+        ...
+
+    @overload
+    def on(self, signal: Literal["message"], callback: Callable[[int, str, Optional[bytes]], None]) -> None:
```
Comment (oriori1703) [inline frida/core.py:1123]:
```suggestion
    def on(self, signal: Literal["message"], callback: Callable[[int, str, Optional[bytes]], Any]) -> None:
```

File: frida/core.py [hunk 6]
```diff
@@ -1161,6 +1201,22 @@ def watch(
         _filter_missing_kwargs(kwargs)
         return self._impl.watch(entrypoint, **kwargs)
 
+    @overload
+    def on(self, signal: Literal["starting" | "finished"], callback: Callable[[], None]) -> None:
```
Comment (oriori1703) [inline frida/core.py:1205]:
```suggestion
    def on(self, signal: Literal["starting", "finished"], callback: Callable[[], Any]) -> None:
```

File: frida/core.py [hunk 7]
```diff
@@ -1161,6 +1201,22 @@ def watch(
...
+        ...
+
+    @overload
+    def on(self, signal: Literal["output"], callback: Callable[[str], None]) -> None:
```
Comment (oriori1703) [inline frida/core.py:1209]:
```suggestion
    def on(self, signal: Literal["output"], callback: Callable[[str], Any]) -> None:
```

File: frida/core.py [hunk 8]
```diff
@@ -1161,6 +1201,22 @@ def watch(
...
+        ...
+
+    @overload
+    def on(self, signal: Literal["diagnostics"], callback: Callable[[Any], None]) -> None:
```
Comment (oriori1703) [inline frida/core.py:1213]:
```suggestion
    def on(self, signal: Literal["diagnostics"], callback: Callable[[Any], Any]) -> None:
```

**Review by oriori1703 — CHANGES_REQUESTED** [review]

File: frida/core.py (hunk 6 above)
Comment (oriori1703) [inline frida/core.py:1205]:
```suggestion
    def on(self, signal: Literal["starting", "finished"], callback: Callable[[], None]) -> None:
```

**Comment by yotamN** [comment]
Bus is the only class that I haven't add a specific handler, I don't think it's a blocker as it's already vastly better.

File: setup.py [hunk 9]
```diff
@@ -108,7 +108,7 @@ def build_extension(self, ext):
         author="Frida Developers",
         author_email="oleavr@frida.re",
         url="https://frida.re",
-        install_requires=["setuptools", "typing_extensions; python_version<'3.8'"],
+        install_requires=["setuptools", "typing_extensions; python_version<'3.11'"],
```
Comment (oriori1703) [inline setup.py:111]: This is not really related to this MR, but setuptools should not be defined as a dependency, because it is only needed during build time.

Instead it should be specified in the `pyproject.toml` according [PEP 517](https://peps.python.org/pep-0517/) and [PEP 518](https://peps.python.org/pep-0518/)

i.e.  :
```toml
[build-system]
requires = ["setuptools"]
build-backend = "setuptools.build_meta"
```

**Review by oriori1703 — CHANGES_REQUESTED** [review]

File: frida/core.py [hunk 10]
```diff
@@ -833,13 +902,69 @@ def get_bus(self) -> Bus:
 
         return self.bus
 
+    @overload
+    def on(self, signal: Literal["spawn_added", "spawn_removed"], callback: Callable[[_frida.Spawn], None]) -> None:
```
[the remark this answers is not in the document]
Comment (yotamN) [reply frida/core.py:906]: yes it does but it's a breaking change, I'm waiting for Ole to decide on when and how to do it.

File: frida/core.py [hunk 11]
```diff
@@ -1087,6 +1244,36 @@ def untag(self, connection_id: int, tag: str) -> None:
...
+    @overload
+    def on(self, signal: Literal["authenticated"], callback: Callable[[int, str], None]) -> None:
+        ...
+
+    @overload
+    def on(self, signal: Literal["subscribe"], callback: Callable[[int], None]) -> None:
+        ...
+
+    @overload
+    def on(self, signal: Literal["message"], callback: Callable[[int, str, Optional[bytes]], None]) -> None:
```
[the remark this answers is not in the document]
Comment (yotamN) [reply frida/core.py:1270]: The nodejs bindings are not the source of truth and it seems they are wrong here.
As can be seen in Frida core, more specifically the `PortalService` class:

```vala
namespace Frida {
	public class PortalService : Object {
                // more signals here...
		public signal void message (uint connection_id, string json, Bytes? data);
```

File: setup.py (hunk 9 above)
Comment (yotamN) [reply setup.py:111]: Would keep in mind but fix in a separate PR
